src/test/kotlin/org/rust/ide/inspections/RsBorrowCheckerInspectionTest.kt

```diff
@@ -0,0 +1,292 @@
+        struct S;
+        impl S {
+            fn foo(&mut self) {}
+        }
+        fn test((x,y): (&S, &mut S)) {
+            <error>x</error>.foo();
+        }
+    """)
+
+    fun `test mutable used at mutable call (pattern)`() = checkByText("""
+        struct S;
+        impl S {
+            fn foo(&mut self) {}
+        }
+        fn test((x,y): (&mut S, &mut S)) {
+            x.foo();
+        }
+    """)
+
+    fun `test mutable used at mutable method definition`() = checkByText("""
+        struct S;
+        impl S {
+            fn test(&mut self) {
+                self.foo();
+            }
+            fn foo(&mut self) {}
+        }
+    """)
+
+    fun `test mutable type should not annotate`() = checkByText("""
+        struct S;
+        impl S {
+            fn foo(&mut self) {}
+        }
+        trait Test {
+            fn test(self);
+        }
+        impl<'a> Test for &'a mut S {
+            fn test(self) {
+                self.foo();
+            }
+        }
+    """)
+
+    fun `test immutable used at mutable method definition`() = checkByText("""
+        struct S;
+        impl S {
+            fn test(&self) {
+                <error descr="Cannot borrow immutable local variable `self` as mutable">self</error>.foo();
+            }
+            fn foo(&mut self) {}
+        }
+    """)
+
+    fun `test immutable used at reference mutable function definition`() = checkByText("""
+        struct S;
+        impl S {
+            fn foo(&mut self) {}
+        }
+        fn test(test: &S) {
+            <error>test</error>.foo();
+        }
+    """)
+
+    fun `test mutable used at reference mutable function definition`() = checkByText("""
+        struct S;
+        impl S {
+            fn foo(&mut self) {}
+        }
+        fn test(test: &mut S) {
+            test.foo();
+        }
+    """)
+
+    fun `test immutable used at reassign`() = checkByText("""
+        fn main() {
+            let a;
+            if true {
+                a = 10;
+            } else {
+                a = 20;
+            }
+            a = 5;//FIXME(farodin91): this line should fail
+        }
+    """)
+
+    fun `test mutable used at reassign`() = checkByText("""
+        fn main() {
+            let mut x;
+            x = 3;
+            x = 5;
+        }
+    """)
+
+    fun `test let some from mutable reference`() = checkByText("""
+        fn foo(optional: Option<&mut Vec<String>>) {
+            if let Some(x) = optional {
+                *x = "str".to_string();
+            }
+        }
+    """)
+
+    fun `test simple enum variant is treated as mutable`() = checkByText("""
+        enum Foo { FOO }
+        fn foo (f: &mut Foo) {}
+        fn bar () {
+            foo(&mut Foo::FOO);     // Must not be highlighted
+        }
+    """)
+
+    @Ignore
+    fun `test fix let at method call (self)`() = checkFixByText("Make `test` mutable", """
```

Comment:
The most interesting fact is that all `@Ignore` tests actually do work in this PR :)